*
*  Redistribution and use in source and binary forms, with or without
*  modification, are permitted provided that the following conditions are met:
*  1. Redistributions of source code must retain the above copyright
*     notice, this list of conditions and the following disclaimer.
*  2. Redistributions in binary form must reproduce the above copyright
*     notice, this list of conditions and the following disclaimer in the
*     documentation and/or other materials provided with the distribution.
*  3. Neither the name of the copyright holder nor the
*     names of its contributors may be used to endorse or promote products
*     derived from this software without specific prior written permission.
*
*  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
*  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
*  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
*  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
*  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
*  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
*  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
*  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
*  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
*  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
*  POSSIBILITY OF SUCH DAMAGE.
*/

/*
* Header for patch information.
* Description           : Add command to generate one time passcode.
* Version               : v001
* Mandatory/Optional    : Mandatory
* Impact if not applied : It requires for thread v1.4 12.3 test case.
*/

Upstream-Status: Inappropriate [platform specific changes]

Upstream-Status: Inappropriate  [platform specific changes]

diff --git a/src/posix/main.c b/src/posix/main.c
index e58f5f142..c320ab8a7 100644
--- a/src/posix/main.c
+++ b/src/posix/main.c
@@ -78,6 +78,7 @@ extern otError ProcessGetSetCcaCfg(void *aContext, uint8_t aArgsLength, char *aA
 extern otError ProcessGetFwVersion(void *aContext, uint8_t aArgsLength, char *aArgs[]);
 extern otError ProcessGetSetSpiFrequency(void *aContext, uint8_t aArgsLength, char *aArgs[]);
 extern otError ProcessGetSetIRThreshold(void *aContext, uint8_t aArgsLength, char *aArgs[]);
+extern otError ProcessGenerateOneTimePasscode(void *aContext, uint8_t aArgsLength, char *aArgs[]);
 
 /**
  * Initializes NCP app.
@@ -374,7 +375,8 @@ static const otCliCommand kCommands[] = {
     {"ccacfg", ProcessGetSetCcaCfg}, //=> Set/Get CCA configuration for 15.4 CCA Before Tx operation
     {"fwversion", ProcessGetFwVersion}, //=> Get firmware version for 15.4
     {"spifreq", ProcessGetSetSpiFrequency},  //=> Get/Set SPI frequency supported by RCP
-    {"irthold", ProcessGetSetIRThreshold}   //=> OutOfBand Independent Reset Threshold configuration
+    {"irthold", ProcessGetSetIRThreshold},   //=> OutOfBand Independent Reset Threshold configuration
+    {"otpc", ProcessGenerateOneTimePasscode} //=> Generate one time passcode
 };
 
 int main(int argc, char *argv[])
diff --git a/src/posix/platform/daemon.cpp b/src/posix/platform/daemon.cpp
index 3daf139da..cf4a0963b 100644
--- a/src/posix/platform/daemon.cpp
+++ b/src/posix/platform/daemon.cpp
@@ -49,6 +49,10 @@
 #include "common/code_utils.hpp"
 #include "posix/platform/platform-posix.h"
 
+#include <openthread/verhoeff_checksum.h>
+#include <openthread/platform/crypto.h>
+#include <openthread/platform/entropy.h>
+
 #if OPENTHREAD_POSIX_CONFIG_DAEMON_ENABLE
 
 #define OPENTHREAD_POSIX_DAEMON_SOCKET_LOCK OPENTHREAD_POSIX_CONFIG_DAEMON_SOCKET_BASENAME ".lock"
@@ -859,6 +863,30 @@ otError ProcessGetSidePanChannel(void *aContext, uint8_t aArgsLength, char *aArg
     return error;
 }
 
+otError ProcessGenerateOneTimePasscode(void *aContext, uint8_t aArgsLength, char *aArgs[])
+{
+    OT_UNUSED_VARIABLE(aContext);
+    OT_UNUSED_VARIABLE(aArgs);
+    OT_UNUSED_VARIABLE(aArgsLength);
+    otError error = OT_ERROR_NONE;
+    uint8_t sEphemeralKey[10];
+    uint8_t  i     = 0;
+    uint32_t randomResult;
+    char     verhoeffChecksum;
+
+    memset(sEphemeralKey, 0, sizeof(sEphemeralKey));
+
+    VerifyOrExit(otPlatEntropyGet((uint8_t *)&randomResult, sizeof(randomResult)) == OT_ERROR_NONE,
+                 error = OT_ERROR_FAILED);
+    randomResult %= 100000000;
+    i += snprintf((char *)sEphemeralKey, sizeof(sEphemeralKey), "%08u", randomResult);
+    VerifyOrExit(otVerhoeffChecksumCalculate((const char *)sEphemeralKey, &verhoeffChecksum) == OT_ERROR_NONE,
+                 error = OT_ERROR_FAILED);
+    i += snprintf((char *)&sEphemeralKey[i], sizeof(sEphemeralKey) - i, "%c", verhoeffChecksum);
+    otCliOutputFormat("%s\r\n",sEphemeralKey);
+exit:
+    return error;
+}
 static const otCliCommand kCommands[] = {
     {"ircfg", ProcessIRConfig},    //=> OutOfBand Independent Reset Configuration ircfg <1> means OOB mode
     {"ircmd", ProcessIRCmd},       //=> InBand Independent Reset command
@@ -869,7 +897,8 @@ static const otCliCommand kCommands[] = {
     {"fwversion", ProcessGetFwVersion}, //=> Get firmware version for 15.4
     {"spifreq", ProcessGetSetSpiFrequency},  //=> Get/Set SPI frequency supported by RCP
     {"irthold", ProcessGetSetIRThreshold},   //=> OutOfBand Independent Reset Threshold configuration
-    {"sidepanch", ProcessGetSidePanChannel} //=> Get Side PAN Channel for 15.4
+    {"sidepanch", ProcessGetSidePanChannel}, //=> Get Side PAN Channel for 15.4
+    {"otpc", ProcessGenerateOneTimePasscode} //=> Generate one time passcode
 };
 } //extern "C"
 
